proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+---
+author: Thomas Bagrel
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/0>).
+**After creating the pull request, edit this file again, update the number in
+the link, and delete this bold sentence.**
+
+# Primitives for zero-copy compact regions
+
+This proposal introduces new primops in GHC to enable direct data structure building in compact regions (without temporary storage in the regular garbage-collected heap), which is useful for performance as it can greatly reduce GC cost.
+
+## Motivation
+
+[Compact regions](https://github.com/ezyang/compact) have been supported in GHC since 2017; they provide a way to allocate memory in a region that can be garbage-collected all at once instead of being traversed fully by the GC at each collection cycle. Compact regions are useful in some performance-critical use-cases, especially when dealing with large long-lived datasets.
+
+So far however, there is no way to operate in a compact region directly. One has to first build data structures in the regular GC heap, and then copy them to a compact region.
+
+I propose a new compact region RTS primitive, [`compactAddHollow#`](https://gitlab.haskell.org/tweag/ghc/-/compare/master...tbagrel1%2Fdps-compact-regions-prims?from_project_id=1#451725cc4e5d443a3b7c2adcdf224840f953b7e2_3537_3537), that allocates data constructor heap objects directly into a compact region, without setting values for their fields yet. `compactAddHollow#` lets us build structures in an _outside-in_ fashion into a region, bypassing entirely the need for a copy from the GC heap.
```

Comment:
You should probably be explicit, here, that you are proposing more than just `compactAddHollow#`.